rclcpp/src/rclcpp/publisher_base.cpp

```diff
@@ -311,6 +329,17 @@ PublisherBase::default_incompatible_qos_callback(
     policy_name.c_str());
 }
 
+void
+PublisherBase::default_incompatible_type_callback(
+  rclcpp::IncompatibleTypeInfo & event) const
+{
+  (void)event;
+
+  RCLCPP_WARN(
+    rclcpp::get_logger(rcl_node_get_logger_name(rcl_node_handle_.get())),
+    "Incompatible type on topic '%s', no messages will be sent to it.", get_topic_name());
```

Comment:
I want to know the behavior of this event.

If any incompatible type is detected in the topic, all endpoints can receive this event and stop sending / receiving messages? from this warning i would expect that behavior in DDS.

or only incompatible type endpoint will receive this event? and then the question is how we can decide on who is incompatible in the topic? Is that `1st comes 1st served` policy?

Some example conditions,

- publisher(type A) / subscription(type A) already communicating. then publisher/subscription(type B) joins and type B endpoints can only receive this event? or publisher(type A) / subscription(type A) also receives the event? and publisher(type A) / subscription(type A) also stops communicating?
- publisher(type A) exists on the topic. then subscription(type B) joins. subscription(type B) will receive the event? and what if after that publisher(type B) joins?

Reply:
as event notification schema, i believe it is really clear from https://github.com/ros2/rmw/pull/339#discussion_r1085538475.

but (just checking) i think message communication will keep sending and receiving message if rmw implementation can do so only with compatible topic types? i think this is important behavior to know for user application.

Reply:
Oh, I see.  These are slightly different questions, sorry I didn't read them closer.

OK, let me go through them:

> publisher(type A) / subscription(type A) already communicating. then publisher/subscription(type B) joins and type B endpoints can only receive this event? or publisher(type A) / subscription(type A) also receives the event? and publisher(type A) / subscription(type A) also stops communicating?

In this case, if a publisher(type B) joins, then publisher(type B) and subscription(type A) both get an event.  publisher(type A) does *not* get an event.  If a subscription(type B) joins, then subscription(type B) and publisher(type A) both get an event.  In all cases, publisher(type A) and subscription(type A) keep communicating.

> * publisher(type A) exists on the topic. then subscription(type B) joins. subscription(type B) will receive the event? and what if after that publisher(type B) joins?

In this case, when subscription(type B) joins, both publisher(type A) and subscription(type B) get the event, and no communication happens.  When publisher(type B) joins, then subscription(type B) starts receiving data from publisher(type B) only.

Does that make sense?

Reply:
it does. even inconsistent topic event comes up, endpoints can communicate each other with compatible type. I just wanted to check this just in case, thank you so much for the clarification.